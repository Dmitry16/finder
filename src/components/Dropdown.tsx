import React, { Fragment, useState, useEffect } from 'react';
import styled from 'styled-components';
import ListItem from './ListItem';
import { findMatch } from '../utils';
import { data } from '../data';

const Input = styled.input`
  margin: 1em;
  padding: 0 0.5em;
  height: 1.5em;
  font-size: 1em;
  background: khaki;
  color: blue;
  border: 2px solid green;
  border-radius: 3px;
`
const List = styled.ul`
  margin: 0;
  text-align: center;
  list-style-type: circle;
`
interface Props {
  listBlockHeight: number;
  passedValue: string;
  clickHandler: (value: string) => void;
}

const Dropdown: React.FC<Props> = ({ passedValue, clickHandler, listBlockHeight }) => {

  const [state, setState] = useState({
    inputValue: passedValue,
    displayMatchesList: false,
    matches: ['']
  });
  // whenever a new value is passed from the parent
  // (after a dropdown option had been selected) it's been displayed as an input value
  useEffect(() => {
    setState({
      ...state,
      inputValue: passedValue,
    })
  }, [passedValue]);
  // handler to process a change event (typing or deleting) in the input
  const changeHandler = (event: any) => {
    const { value } = event.target;
    setState({
      ...state,
      displayMatchesList: !!value,
      inputValue: value,
      matches: findMatch(value, data)
    });
  }

  const { displayMatchesList, matches} = state;

  return (
    <Fragment>
      Type in your search term then select an option from the list:
      <Input type='text' name='dropdown' value={state.inputValue} onChange={changeHandler} />
      <List>
        {
          displayMatchesList && matches.length &&
            matches.filter((_, ind) => ind <= listBlockHeight-1)
              .map((match: string, ind: number) =>
                <ListItem key={ind} clickHandler={clickHandler} match={match} />)
        }
      </List>
    </Fragment>
  );
}

export default Dropdown;
