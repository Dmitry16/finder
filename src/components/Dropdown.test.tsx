import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

import Dropdown from './Dropdown';

describe('<Dropdown />', () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = mount(<Dropdown
      passedValue={''}
      clickHandler={() => {}}
      listBlockHeight={5}
    />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Dropdown 
      passedValue={''}
      clickHandler={() => {}}
      listBlockHeight={5}
    />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders determined layout', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('has an input', () => {
    expect(wrapper.find('input').length).toBe(1);
  });
  // console.log(wrapper.debug());
});

