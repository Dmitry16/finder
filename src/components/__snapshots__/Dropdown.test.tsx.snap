// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders determined layout 1`] = `"Type in your search term then select an option from the list:<input type=\\"text\\" name=\\"dropdown\\" class=\\"sc-bwzfXH dKDNGw\\" value=\\"\\"><ul class=\\"sc-htpNat hXEaPD\\"></ul>"`;
